### PR TITLE
feat(platform-rivet): add rivet as a platform host

### PR DESCRIPTION
**Pull request by prisis**
## Summary

Adds **`@lunora/platform-rivet`**, a third `@lunora/platform` host implementing `ShardHost`, `SocketHost`, `ShardDirectory`, `ShardKvStore` and `SchedulerHost` over [Rivet Actors](https://rivet.dev/actors) (`rivetkit`), plus `RIVET_CAPABILITIES` in `@lunora/platform` and its registration in `@lunora/codegen`.

Touched packages: `@lunora/platform-rivet` (new), `@lunora/platform`, `@lunora/codegen`.

A Rivet Actor is the closest primitive to a Durable Object outside Cloudflare — one addressable, single-writer instance per key, with its own SQLite database, its own durable schedules, and a sleep/wake lifecycle. Because plan 114 already split contracts / engine / hosts, the port touched **none** of `@lunora/do`, `@lunora/runtime` or `@lunora/server`; it is one new package implementing five interfaces. That supersedes `plans/multi-platform-portability-assessment.md`'s verdict ("porting would mean rewriting…"), which predates the split.

Several rows are rated `native` where `@lunora/platform-node`'s read `emulated`:

| Contract | Rivet primitive | Rating |
| --- | --- | --- |
| Sharded state | one actor per shard key | native |
| Shard alarms | `c.schedule.at` (wakes a sleeping actor) | native |
| Hibernated WebSockets | `options.canHibernateWebSocket` | native |
| Scheduler, incl. **runtime cron** | `c.schedule` + `c.cron.set` | native |
| Placement at create time | `getOrCreate(key, { createInRegion })` | emulated¹ |
| Local SQL | `c.db`, behind a synchronous working copy | emulated |

¹ Native in Rivet, but region slugs are deployment-defined, so the caller supplies the mapping; without one the hint is dropped rather than guessed.

This is the first host to implement the optional `SchedulerHost.cron` from a platform primitive rather than its own table — Rivet registers crons at runtime, which Cloudflare cannot.

### The one structural mismatch

`ShardSqlExec.exec` is **synchronous** (the engine walks cursors without awaiting) while every Rivet SQLite entry point is a promise, because the actor reaches its database through the runtime. The contract anticipates this ("async-backed with a sync facade"), and `src/rivet-shard-state.ts` is that facade: a `better-sqlite3` working copy in the actor's memory answers every read, hydrated on wake from a snapshot and serialized back at each commit.

A snapshot rather than a statement journal, because replaying `random()`/`datetime('now')` reconstructs a *different* database. The cost is real and documented: a commit is O(database size), Rivet caps an actor at 10 GiB shared between SQLite and KV, and it stores SQLite *through* that KV layer — so this is a small-shard strategy that `.shardBy()` keeps in range. That is why `localSql` is `emulated`.

### `crossShardFanout` is `unsupported`, deliberately

The query coordinator needs to **enumerate** shard keys; the RivetKit client only **addresses** them. A fan-out query would answer from no shards instead of all of them — a wrong answer, not a slow one — so it fails codegen's gate rather than shipping a silent hole.

## Linked issues

None.

## Test plan

- [x] `pnpm --filter "@lunora/platform-rivet" run lint:types` passes
- [x] `pnpm --filter "@lunora/platform-rivet" run lint:eslint` passes (`--max-warnings=0`)
- [x] `pnpm --filter "@lunora/platform-rivet" run test:coverage` — 56 tests, all green, above the package's floors
- [x] `pnpm --filter "@lunora/platform" run test` — 45 tests green
- [x] `pnpm --filter "@lunora/codegen" exec vitest run __tests__/platform-target.test.ts` — 13 tests green
- [x] `pnpm --filter "@lunora/config" exec vitest run __tests__/project-config.test.ts __tests__/driver-registry.test.ts` — the `deployTargetIds ⊆ matrixIds` invariant still holds
- [x] `pnpm run api:check` — 48 snapshots match (new `api-snapshots/platform-rivet.api.md`, experimental tier)
- [x] `pnpm run lint:prettier`, `pnpm run lint:package-json`, `node scripts/check-generated-files.mjs`, `node scripts/check-roadmap-tiers.js`, `node scripts/check-agents-md-packages.js`, `pnpm run lint:node-capabilities-docs`

Reviewers should re-run the full `pnpm run test` and the workerd pool suites — this branch was verified per-package, not with a whole-repo run.

The 32-leg `@lunora/platform/conformance` TCK passes with **no skips**, the same suite that pins Cloudflare and Node.

## Checklist

- [x] Commit messages follow Conventional Commits
- [x] Added or updated tests covering the change
- [x] Updated relevant docs (`packages/platform-rivet/README.md` + `docs/index.mdx`, `packages/platform/docs/index.mdx`, `AGENTS.md`, `ROADMAP.md`, `plans/rivet-host-findings.md`)
- [x] No `package.json` files in `packages/*` modified outside the touched packages
- [x] New package has `project.json` with `type:package` and `category:platform`
- [ ] Migration impact: none — this is additive; no existing target's behaviour changes

## Notes for reviewers

**Two bugs were found by a self-audit after the first commit; both fixes are in this branch.** They are worth looking at first because they are the kind that pass every linter:

1. **`RIVET_CAPABILITIES` was inert** (`f4b9d91`). It shipped complete and unregistered in codegen's `PLATFORM_MATRICES`, and an unregistered id makes `gatePlatformFeatures` return `platform_unknown_target` and *leave the usage set untouched* — so `"target": "rivet"` would have emitted the full Cloudflare-shaped surface with no diagnostic until runtime. Registered, plus a coverage guard driven off `@lunora/platform`'s own exports so the next host is covered the day its matrix lands.
2. **`keyValueStore` was rated `native`** describing `c.kv`, but the host deliberately avoids it (deprecated on every member) and writes a `_lunora_kv` table instead. The module docstring, README and findings doc all said `emulated`; only the matrix — the one thing codegen reads — was wrong.

**Please sanity-check these judgement calls:**

- **The snapshot facade.** It is correct and expensive. `plans/rivet-host-findings.md` §2 and §7 argue for revisiting it (incremental/dirty-page snapshot, BLOB instead of base64) and ask the broader question: should `ShardHost` grow an async-first read path now that a second host has hit the sync/async boundary?
- **Two storage mechanisms on purpose** (findings §3): synchronous contracts go through the working copy; asynchronous ones (`ShardKvStore`, the scheduler's job table) write straight through to `c.db`. Split by contract shape, not convenience.
- **`jurisdiction()` is deliberately absent** from the directory. A jurisdiction is a hard residency constraint; Rivet's region selection is best-effort placement. Mapping one onto the other would turn "this data may not leave the EU" into "we tried", so callers fail closed as the contract instructs.
- **`scripts/check-node-capabilities-docs.js` now covers every host's capability page**, not just Node's. The filename is unchanged on purpose — a CI job, a path filter and a root script name key off it.

**Two caveats that bound what this proves:**

1. The TCK runs against an **in-memory actor double**, not a live Rivet engine. It proves the adapters satisfy the contracts given a context that behaves the way Rivet documents; it does not prove Rivet behaves that way. Open items (schedule delivery across a real sleep, napi-vs-wasm parameter serde, hibernation wake ordering) are listed in findings §6.
2. `src/rivet-context.ts` is a hand-written narrowing of `rivetkit@2.3.10`'s types, drift-checked against **copies** of the upstream declarations rather than a real dependency — pulling `rivetkit` in for types alone would add the napi + wasm binaries, `drizzle-orm`, `hono` and `pino` to every install. It catches a projection that was always wrong; it catches drift only when someone refreshes it.

**Deferred, not forgotten:** no `lunora dev --target rivet`, no `@lunora/config` deploy driver (so `deployTargetIds()` correctly excludes `rivet`), no `.global()` backend, no queue/object-storage/AI bindings. Rivet's own `rivetkit/workflow` SDK and container runner are the obvious next targets. The package is gated at the **experimental** API tier and carries no SemVer promise.

## Contributor License Agreement

> By submitting this pull request, I confirm that my contribution is made under the terms of the project's license and that you can use, modify, copy, and redistribute this contribution under those terms.

---
_Generated by [Claude Code](https://claude.ai/code/session_01AveZDzzUfBmDQaTFeEcdAa)_

## Summary by CodeRabbit

* **New Features**
  * Added an experimental Rivet host package supporting actor-backed state, alarms, scheduling, WebSockets, key-value storage, and shard routing.
  * Added Rivet capability support, platform listings, and conformance validation.
  * Added shared scheduling retry behavior across supported hosts.

* **Bug Fixes**
  * Improved key-value prefix queries to correctly handle special characters and maintain predictable ordering.

* **Documentation**
  * Added Rivet setup guidance, supported features, limitations, configuration requirements, and capability details.
  * Expanded capability documentation checks to cover multiple host packages.

### CODE REVIEW

File: packages/platform-node/src/node-kv-store.ts [hunk 1]
```diff
@@ -43,7 +54,10 @@ export const createNodeShardKvStore = (database: Database.Database): ShardKvStor
         "INSERT INTO _lunora_kv (key, value) VALUES (?, ?) ON CONFLICT (key) DO UPDATE SET value = excluded.value",
     );
     const deleteStatement = database.prepare<[string]>("DELETE FROM _lunora_kv WHERE key = ?");
-    const scanStatement = database.prepare<[], { key: string; value: Buffer }>("SELECT key, value FROM _lunora_kv");
+    const scanStatement = database.prepare<[], { key: string; value: Buffer }>("SELECT key, value FROM _lunora_kv ORDER BY key");
+    const prefixStatement = database.prepare<[string], { key: string; value: Buffer }>(
+        String.raw`SELECT key, value FROM _lunora_kv WHERE key LIKE ? ESCAPE '\' ORDER BY key`,
+    );
```
[the remark this answers is not in the document]
Comment (prisis) [reply packages/platform-node/src/node-kv-store.ts:66]: Fixed in 642ba8e59. Kept the `LIKE` as the index-friendly prefilter and made `startsWith` the authority over what `list` returns — same over-match the `_`/`%` escaping already guards, reached by a second route the escaping never touched. Added a regression test over ASCII case variants; it fails against the previous code.

File: packages/platform-rivet/src/rivet-socket-host.ts [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+                handle: undefined as unknown as SocketHandle,
+                id,
+                raw: undefined,
+                tags: new Set(row === undefined ? [] : (JSON.parse(row.tags) as string[])),
+            };
+
+            runtimeSockets.set(id, record);
+
+            return createHandle(record, undefined);
+        },
```
[the remark this answers is not in the document]
Comment (prisis) [reply packages/platform-rivet/src/rivet-socket-host.ts]: Fixed in 642ba8e59 — the analysis was exactly right, including that the state is marked dirty and the snapshot rewritten, so nothing surfaces until the next recycle. `materialize` now upserts the row when there was none, matching what `accept` does.

File: packages/platform-rivet/packem.config.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import type { BuildConfig } from "@visulima/packem/config";
+import { defineConfig } from "@visulima/packem/config";
+import transformer from "@visulima/packem/transformer/esbuild";
+
+// eslint-disable-next-line import/no-unused-modules -- consumed by packem CLI
+export default defineConfig({
+    runtime: "node",
+    failOnWarn: false,
```
[the remark this answers is not in the document]
Comment (prisis) [reply packages/platform-rivet/packem.config.ts:8]: Not changing this one. `failOnWarn: false` is the repo-wide packem convention — `packages/platform-node`, `packages/platform` and `packages/do` all carry the identical line, so this is not the branch relaxing a gate but a new package following the existing one. Flipping it here alone would make rivet the only package built under different rules. Worth raising repo-wide if the setting is wrong, but that is not a change a host-spike PR should make unilaterally.

File: packages/platform-rivet/package.json [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    "devDependencies": {
+        "@types/better-sqlite3": "catalog:tooling",
+        "@types/node": "catalog:types",
+        "@visulima/packem": "catalog:build",
+        "@vitest/coverage-v8": "catalog:test",
+        "esbuild": "catalog:build",
+        "fallow": "catalog:lint",
+        "typescript": "catalog:tsc",
+        "vitest": "catalog:test"
+    },
```
[the remark this answers is not in the document]
Comment (prisis) [reply packages/platform-rivet/package.json:77]: Not adding it. Zero of the 57 packages in `packages/*` declare `@anolilab/eslint-config` — it is a root devDependency and every package's `eslint.config.js` imports it from there. Adding it here would make rivet the only package that declares it, against the convention rather than toward it.

File: packages/platform-rivet/src/rivet-scheduler-host.ts [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+                retry?.backoffMultiplier ?? DEFAULT_RETRY.backoffMultiplier,
+                retry?.maxDelayMs ?? DEFAULT_RETRY.maxDelayMs,
+                // eslint-disable-next-line unicorn/no-null -- SQL NULL for an absent routing hint; the column is nullable and `undefined` is not bindable
+                scheduleOptions?.shardKey ?? null,
+            );
+
+            await arm(id, scheduledFor);
+
+            return { id, scheduledFor };
+        },
```
[the remark this answers is not in the document]
Comment (prisis) [reply packages/platform-rivet/src/rivet-scheduler-host.ts:310]: The gap is real but the premise is not: rivet does NOT store . Its jobs table (, the CREATE TABLE) has no such column, and neither host persists or forwards the value — so  is documented as "forwarded to the worker so the call lands on the right shard" while nothing anywhere reads it. That is worse than dropping it late, and it is the same inert-surface class AGENTS.md warns about.

Leaving this open rather than fixing it here. Adding `shardKey` to `ScheduledJob` changes `@lunora/platform`'s public surface, and the repo's own rule — the one cited when `ShardHost.asyncSql` was deleted for having no readers — is that such a field lands with its consumer, an implementation on every host, and a TCK leg in one change. Doing that inside a PR whose subject is adding a host would be the wrong shape, and half-doing it would recreate exactly the defect you found.

[the remark this answers is not in the document]
Comment (prisis) [reply packages/platform-rivet/src/rivet-scheduler-host.ts:310]: Correcting my reply above — the backticks were eaten by my shell, so the first paragraph lost its code spans. Restating it properly:

The gap is real but the premise is not: rivet does **not** store `shard_key`. Its jobs table (the `CREATE TABLE` in `rivet-scheduler-host.ts`) has no such column, and neither host persists or forwards the value — so `ScheduleOptions.shardKey` is documented as "routing hint forwarded to the worker so the call lands on the right shard" while nothing anywhere reads it. That is worse than dropping it late, and it is the same inert-surface class `AGENTS.md` warns about.

The rest of my reply stands: leaving this open rather than fixing it here, because adding `shardKey` to `ScheduledJob` changes `@lunora/platform`'s public surface, and the repo's own rule — the one cited when `ShardHost.asyncSql` was deleted for having no readers — is that such a field lands with its consumer, an implementation on every host, and a TCK leg in one change.

File: packages/platform-rivet/src/rivet-shard-host.ts [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+        const run = async (): Promise<T> => runTransaction(function_);
+        const started = transactionTail.then(run, run);
+
+        transactionTail = started.then(
+            () => undefined,
+            () => undefined,
+        );
+
+        return started;
+    };
```
[the remark this answers is not in the document]
Comment (prisis) [reply packages/platform-rivet/src/rivet-shard-host.ts:285]: Leaving this one open deliberately.

The analysis reads correct to me — `runSerialized` and `transaction` really are separate tails, `flush` has no tail of its own, and raw `BEGIN`/`COMMIT` on one shared `better-sqlite3` connection plus a single `id = 0` snapshot row is a genuine race. I am not fixing it in this pass because a shared-mutex-with-reentrancy change to a host's concurrency boundaries is not something to land on inference: the composition you name (`runSerialized(() => transaction(work))`) has to stay deadlock-free, and getting that subtly wrong is worse than the race it replaces, because it would hang rather than corrupt.

It also wants a test that actually interleaves two boundaries and asserts the snapshot is the newest one — otherwise the fix is a claim nothing enforces. That is a focused piece of work, and it deserves its own change rather than being folded into the tail of a review pass.

File: packages/platform-rivet/src/rivet-shard-state.ts [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+            // would drop the very writes a failed flush still owes, and the
+            // next boundary would find nothing to retry.
+            dirty = false;
+        }
+
+        // No guard here, and none needed: nothing ever opens a transaction on
+        // the registry copy — that is what it is separate for.
+        if (registryDirty) {
+            await persist(REGISTRY_SLOT, registry);
+            registryDirty = false;
```
[the remark this answers is not in the document]
Comment (prisis) [reply packages/platform-rivet/src/rivet-shard-state.ts]: Fixed in 86d8df66e. Each working copy now carries a revision the markers bump; `flush` captures it before awaiting and clears the flag only if it has not moved, so a write that lands inside the await window stays owed and the next boundary re-flushes. Conservative in the safe direction — worst case is one redundant snapshot. Regression test drives that interleaving and fails against the previous code.

File: packages/platform-rivet/src/rivet-shard-host.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (prisis) [reply packages/platform-rivet/src/rivet-shard-host.ts]: Confirmed and fixed in d094e9639. Both halves were real; one detail of the diagnosis was not.

**Not reproducible as stated:** concurrent boundaries could *not* nest a raw `BEGIN`. Only `transaction` issued one, and `transactionTail` already serialized every `transaction` against every other, so `BEGIN` depth never exceeded 1.

**Real, and worse than a nested `BEGIN`:** `runSerialized` writes without a `BEGIN` of its own, so when it overlapped a bare `transaction()` its rows landed *inside* that transaction. Its `flush()` then saw `database.inTransaction` and skipped the snapshot (it has to — `serialize()` captures uncommitted rows), `runSerialized` **resolved**, and the transaction's `ROLLBACK` destroyed a write already reported as durable. Gone from the working copy and from Rivet's SQLite, nothing having failed. New test `keeps a serialized boundary's write out of a concurrent transaction` reads `[]` instead of `[{id: 2}]` on the pre-fix host.

**Also real, and not covered by the existing revision guard:** `flush` had no tail. `RivetPlatform.flush` is public and `close()` calls it from the sleep path, which Rivet does not serialize against an in-flight action. Two flushes serialize their bytes up front and then race on `db.execute`, where Rivet promises no ordering — so the *older* snapshot can land last while the newer flush has already cleared `dirty` for bytes no longer in the row. Nothing owes a retry; the next wake hydrates a shard missing its last write.

**The fix.** One `withBoundary` lock shared by `runSerialized` and `transaction`, plus a `flushTail` in `rivet-shard-state.ts`.

Re-entrancy is `AsyncLocalStorage`, not a held-boolean. A boolean reads `true` for an unrelated second top-level `transaction()` too and *would* nest a raw `BEGIN` — the bug your comment described, introduced by the naive fix for it. ALS's store follows one call's own await chain without leaking into a sibling, so only a genuinely nested boundary bypasses the queue. That keeps `ShardRunner.runInTransaction`'s `runSerialized(() => transaction(work))` and `ShardDO.fetch`'s outer `runSerialized` span deadlock-free. Same model `@lunora/do`'s idempotency suite verified against real workerd.

Two notes beyond the finding:
- `runSerialized(() => runSerialized(...))` — the shape `ShardDO.fetch` produces — **deadlocked outright** on the pre-fix host, since the inner enqueue waited on the outer's own tail. The new test for it times out at 5s pre-fix and passes now.
- `dirty` being cleared by a later flush was already closed by the revision counter in 86d8df66e. The remaining gap was purely the out-of-order `persist`, which the tail closes.

**Tests:** 4 interleaving tests in `__tests__/rivet-shard-host.concurrency.test.ts`. Verified against reverted sources: 3 fail pre-fix (one by deadlocking), the 4th (`never nests a raw BEGIN when three boundaries overlap`) passes both ways and is there as a regression guard on the re-entrancy check. The existing `stays dirty when a write lands while the snapshot is being persisted` test now holds its durable write open with a gate — flushes queue before they serialize, so a same-tick mark is legitimately part of the snapshot; it still pins the same property on both versions.

**Gates:** package test 72 passed / 1 skipped; `lint:affected:types` 77 projects; `lint:affected:eslint` 118 projects; `api:check` matches all 49 snapshots; `lint:package-json`; `build` green (`--isolatedDeclarations` d.ts emit). No `PlatformCapabilities` change — host-internal plumbing, no new `ctx.*` surface. Not covered by `test:workerd`: `platform-rivet` has no workerd project and `better-sqlite3` cannot load there.
